libraries/AP_Networking/AP_Networking_port.cpp

```diff
@@ -0,0 +1,181 @@
+
+extern const AP_HAL::HAL& hal;
+
+const AP_Param::GroupInfo AP_Networking::Port::var_info[] = {
+    // @Param: TYPE
+    // @DisplayName: Port type
+    // @Description: Port type
+    // @Values: 0:Disabled, 1:UDP client
+    // @RebootRequired: True
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("TYPE", 1,  AP_Networking::Port, type, 0, AP_PARAM_FLAG_ENABLE),
+
+    // @Param: PROTOCOL
+    // @DisplayName: protocol
+    // @Description: protocol
+    // @Values: -1:None, 1:MAVLink1, 2:MAVLink2, 3:Frsky D, 4:Frsky SPort, 5:GPS, 7:Alexmos Gimbal Serial, 8:Gimbal, 9:Rangefinder, 10:FrSky SPort Passthrough (OpenTX), 11:Lidar360, 13:Beacon, 14:Volz servo out, 15:SBus servo out, 16:ESC Telemetry, 17:Devo Telemetry, 18:OpticalFlow, 19:RobotisServo, 20:NMEA Output, 21:WindVane, 22:SLCAN, 23:RCIN, 24:EFI Serial, 25:LTM, 26:RunCam, 27:HottTelem, 28:Scripting, 29:Crossfire VTX, 30:Generator, 31:Winch, 32:MSP, 33:DJI FPV, 34:AirSpeed, 35:ADSB, 36:AHRS, 37:SmartAudio, 38:FETtecOneWire, 39:Torqeedo, 40:AIS, 41:CoDevESC, 42:DisplayPort, 43:MAVLink High Latency, 44:IRC Tramp
+    // @RebootRequired: True
+    // @User: Advanced
+    AP_GROUPINFO("PROTOCOL", 2,  AP_Networking::Port, state.protocol, 0),
+
+    // @Group: IP
+    // @Path: AP_Networking_address.cpp
+    AP_SUBGROUPINFO(ip, "IP", 3,  AP_Networking::Port, AP_Networking_IPV4),
+
+    // @Param: PORT
+    // @DisplayName: Port number
+    // @Description: Port number
+    // @Range: 0 65535
+    // @RebootRequired: True
+    // @User: Advanced
+    AP_GROUPINFO("PORT", 4,  AP_Networking::Port, port, 0),
+    
+    AP_GROUPEND
+};
+
+/*
+  initialise mapped network ports
+ */
+void AP_Networking::ports_init(void)
+{
+    // init in reverse order to keep the linked list in
+    // AP_SerialManager in the right order
+    for (int8_t i=ARRAY_SIZE(ports)-1; i>= 0; i--) {
+        auto &p = ports[i];
+        NetworkPortType ptype = (NetworkPortType)p.type;
+        p.state.idx = AP_SERIALMANAGER_NET_PORT_1 + i;
+        switch (ptype) {
+        case NetworkPortType::NONE:
+            break;
+        case NetworkPortType::UDP_CLIENT:
+            p.udp_client_init();
+            break;
+        }
+        if (p.sock != nullptr) {
+            AP::serialmanager().register_port(&p);
+        }
+    }
+}
+
+/*
+  initialise a UDP client
+ */
+void AP_Networking::Port::udp_client_init(void)
+{
+    if (!init_buffers(4096)) {
+        return;
+    }
+    sock = new SocketAPM(true);
+    if (sock == nullptr) {
+        return;
+    }
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_Networking::Port::udp_client_loop, void), "NET", 2048, AP_HAL::Scheduler::PRIORITY_UART, 0)) {
+        AP_BoardConfig::allocation_error("Failed to allocate UDP client thread");
+    }
+}
+
+/*
+  update a UDP client
+ */
+void AP_Networking::Port::udp_client_loop(void)
+{
+    while (!hal.scheduler->is_system_initialized()) {
+        hal.scheduler->delay(100);
+    }
+    hal.scheduler->delay(1000);
```

Comment:
is there any better way to do this more intelligently? All over the place in the code we add delays to ensure we don't send GCS msgs out too early... but SURELY there's some sort of status we can check somewhere to know it's a proper time

Reply:
for what it's worth, on my AP_Periph version of this I had to bump this up a little